tests/unit/test_knob_extractor.py

```diff
@@ -0,0 +1,221 @@
+        "knob1": KnobEntry(madx="knob1.madx", lsa="knob1.lsa", scaling=-1, value=12.43383),
+        "knob2": KnobEntry(madx="knob2.madx", lsa="knob2.lsa", scaling=1, value=-3.0231),
+        "knob3": KnobEntry(madx="knob3.madx", lsa="knob3.lsa", scaling=-1, value=-9.7492),
+    }
+    path = tmp_path / "knobs.txt"
+    time = datetime.now()
+    _write_knobsfile(path, knobs_dict, time=time)
+    read_as_dict, full_text = parse_output_file(path)
+    assert str(time) in full_text
+    assert " mo " in full_text
+    assert " Other Knobs " in full_text
+    assert len(read_as_dict) == len(knobs_dict)
+    for _, entry in knobs_dict.items():
+        assert read_as_dict[entry.madx] == entry.value * entry.scaling
+
+
+@pytest.mark.basic
+def test_time_and_delta():
+    time_str = "2022-06-26T03:00"
+    t1 = _parse_time(time_str)
+
+    assert t1 == datetime(2022, 6, 26, 3, 0, 0)
+
+    # 2 hours earlier
+    t2 = _add_time_delta(t1, "_2h")
+    assert t2 == datetime(2022, 6, 26, 1, 0, 0)
+
+    # 1 week earlier
+    t2 = _add_time_delta(t1, "_1w")
+    assert t2 == datetime(2022, 6, 19, 3, 0, 0)
+
+    # 1 week and 1 hour earlier
+    t2 = _add_time_delta(t1, "_1w1h")
+    assert t2 == datetime(2022, 6, 19, 2, 0, 0)
+
+    # 1 month later
+    t2 = _add_time_delta(t1, "1M")
+    assert t2 == datetime(2022, 7, 26, 3, 0, 0)
+
+    # 20 years later
+    t2 = _add_time_delta(t1, "20Y")
+    assert t2 == datetime(2042, 6, 26, 3, 0, 0)
+
+    t3 = _parse_time(time_str, "20Y")
+    assert t2 == t3
+
+
+@pytest.mark.basic
+def test_timezones():
+    assert (
+            _parse_time("2022-06-26T03:00+02:00")
+            ==
+            datetime(2022, 6, 26, 3, 0, 0, tzinfo=timezone(timedelta(seconds=7200)))
+    )
+
+# CERN Tests -------------------------------------------------------------------
+
+
+@pytest.mark.cern_network
+def test_real(tmp_path, knob_definitions, saved_knob_txt):
+    path_saved, time_saved = saved_knob_txt
+    path = tmp_path / "knobs.txt"
+    main(time=time_saved, output=path, knob_definitions=knob_definitions)
+
+    parsed_output, _ = parse_output_file(path)
+    parsed_saved, _ = parse_output_file(path_saved)
+
+    assert len(parsed_saved) == len(parsed_output)
+    for key in parsed_output.keys():
+        assert parsed_output[key] == parsed_saved[key]
+
+
+# Helper -----------------------------------------------------------------------
+
+
+def parse_output_file(file_path):
+    txt = Path(file_path).read_text()
+    d = {}
+    pattern = re.compile(r"\s*(\S+)\s*:=\s*([^;\s*]+)\s*;")
+    for line in txt.splitlines():
+        line = line.strip()
+        if not line or line.startswith("!"):
+            continue
+
+        match = pattern.match(line)
+        d[match.group(1)] = float(match.group(2))
+    return d, txt
+
+
+@pytest.fixture()
+def knob_definitions():
+    return INPUTS / "knob_definitions.txt"
+
+
+@pytest.fixture()
+def saved_knob_txt():
```

Comment:
Give me a type hiint here at east that is also returns the timestamp. Not clear from the fixture name